### ISSUE TITLE
Wrong target reported when using AutoDependencyTracking and HTTP_PROXY environment variable

### ISSUE THREAD

**Issue by wiktorn** [microsoft/ApplicationInsights-node.js#719]
Follow up after #689 and #692.

The problem still persists, that when using request library together with HTTP_PROXY setting, all dependency requests that go through proxy have target of 443 (this are https requests).

Options object in [HttpDependencyParser._getUrlFromRequestOptions](https://github.com/microsoft/ApplicationInsights-node.js/blob/develop/AutoCollection/HttpDependencyParser.ts#L115) is of following structure:
 
```js
{
  host: "proxy-host",
   port: 3128,
   proxyAuth: null,
   headers: {
     Accept: "application/json",
     host: "destination-host:443",
   },
   method: "CONNECT",
   path: "destination-host:443",
   agent: false,
 }
 ```
 
And after `url.parse` in line 141, the protocol is "destination-host" and host is 443.
 
Anyway - the result of `_getUrlFromRequestOptions` function will be in this case still PROXY host, not the real one.
 
On the other hand, when I debug through `public static trackRequest(client: TelemetryClient, telemetry: Contracts.NodeHttpDependencyTelemetry) {`
 
I see requestParser.url set proxy host, but when callback is called `telemetry.request.on('response', (response: http.ClientResponse) => {` then `requestParser.url` is already set to target host and port, but without protocol. Can't find place, where it's actually changed.

**Comment by hectorhdzg** [comment]
@wiktorn are you asking to add protocol in the target field?, I can see current (code)[https://github.com/microsoft/ApplicationInsights-node.js/blob/develop/AutoCollection/HttpDependencyParser.ts#L57] ignore the protocol and only adds hostname and port, you can see it uses let urlObject = url.parse(this.url); to get the target value, requestParser.url is only added during initialization and never updated after.

**Comment by wiktorn** [comment]
@hectorhdzg In case when HTTP_PROXY is used for https traffic what I'd like to have in target field in AppInsights, is the real hostname. Currently I get only `443` which is not satisfying.

Based on my debugging, I can see two calls to HttpDependencyParser, one with HTTP_PROXY host, the other - with the real endpoint.

First call could be summarized as following test case:
```typescript
        it("should return correct data for HTTPS Proxy CONNECT call", () => {
            (<any>request).agent.protocol = "http:";
            (<any>request)["method"] = "CONNECT";
            (<any>request)["path"] = "destination-host:443";
            let requestOptions = {
                "host": "proxy-host",
                "method": "CONNECT",
                "path": "destination-host:443"
            };

            let parser = new HttpDependencyParser(requestOptions, request);

            response.statusCode = 200;
            
            parser.onResponse(response);
            let dependencyTelemetry = parser.getDependencyTelemetry();
            assert.equal(dependencyTelemetry.target, "destination-host");
        });
```

Though I'm not if this get at all logged into App Insights, or maybe it's ignored somewhere else.

The second call, with real data (and the one that actually fails) looks like that:
```typescript
        it("should return correct data for HTTPS Proxy 2", () => {
            (<any>request).agent.protocol = undefined;
            (<any>request)["method"] = "GET";
            (<any>request)["path"] = "/realpath";
            let requestOptions = {
                "host": "destination-host",
                "method": "GET",
                "path": "/realpath",
                "port": "443"
            };

            let parser = new HttpDependencyParser(requestOptions, request);

            response.statusCode = 200;
            
            parser.onResponse(response);
            let dependencyTelemetry = parser.getDependencyTelemetry();
            assert.equal(dependencyTelemetry.target, "destination-host");
        });
```

And for this instance, I see that the target is set to `443`, exactly what I observe in App Insights and what I'd like to get fixed. Maybe its also worth noting, that in this case, `request.agent` is `TunnelingAgent` (probably [this one](https://github.com/request/tunnel-agent)).

I hope that this will make more clear, what the bug is about.

**Comment by hectorhdzg** [comment]
Thanks for extra details @wiktorn, second test case is failing because protocol is not available and URL parsing will fail [here](https://github.com/microsoft/ApplicationInsights-node.js/blob/develop/AutoCollection/HttpDependencyParser.ts#L49), if protocol is not provided as part of the agent or within the options, the SDK will not be able to determine the protocol [here](https://github.com/microsoft/ApplicationInsights-node.js/blob/develop/AutoCollection/HttpDependencyParser.ts#L166) and will fail to parse the URLs correctly, you can try to send the protocol options or add the property somehow using your Agent or manually call the APIs to instrument your dependency calls and have full control of what data is being sent.

**Comment by wiktorn** [comment]
Does it mean, that there are following options:
1. Fix [requests/tunnel-agent](https://github.com/request/tunnel-agent/) so it will provide protocol information
2. Monkey-patch `tunnel-agent` so it will provide protocol information
3. Manually track all dependencies

As option 1 is probably not viable, as requests is deprecated, does it mean, that AutoDependency tracking does not work for requests library when proxy is used?

It would be worth noting somewhere in documentation then about this limitation.

**Comment by hectorhdzg** [comment]
@wiktorn yes those options sound about right, we don't have direct support for request library but we do support native http and https libraries in node.js and that cover plenty of scenarios but not all of them for multiple libraries that use it, including request library. I will update documentation to reflect this so is 100% clear.

**Comment by wiktorn** [comment]
One question regarding migration to other libraries. As we will then use some other agent (that supports proxy), I've checked the node docs, and cant see that http.agent.protocol is part of the [spec](https://nodejs.org/api/http.html#http_class_http_agent), so I fear, that whichever other library we will use, the above code will still fail, as other agents will also not provide this information.

On the other hand, I that recent versions of node (Added in: v14.5.0, v12.19.0) there is added field - `request.protocol` - maybe this is the place, where Azure Insights should also look. As far as I've tested (on v12.20.1), in both above cases request.protocol is set, though to this suspicious value `https:` (with colon at the end).
